server/data/data-jmap/src/test/java/org/apache/james/jmap/api/preview/MessagePreviewStoreContract.java

```diff
@@ -119,25 +120,37 @@ default void storeShouldBeIdempotent() {
     }
 
     @Test
-    default void concurrentStoreShouldStoreAllRecords() throws Exception {
+    default void concurrentStoreShouldStoreTheLatestValueDespiteOfCollisions() throws Exception {
         int threadCount = 10;
         int stepCount = 100;
-
-        ConcurrentTestRunner.builder()
-            .operation((thread, step) -> {
-                int id = thread * stepCount + step;
-                Mono.from(testee().store(TestMessageId.of(id), Preview.from(String.valueOf(id))))
-                    .block();
+        int repeatCount = 10;
+
+        ConcurrentHashMap<Integer, Integer> indexCounter = new ConcurrentHashMap<>();
+
+        Flux.range(0, stepCount)
+            .repeat(repeatCount)
+            .parallel(threadCount)
+            .runOn(Schedulers.newParallel("preview", threadCount))
+            .flatMap(index -> {
+                Integer counter = indexCounter.compute(index, (id, count) -> {
+                    if (count == null) {
+                        return 0;
+                    }
+                    return count + 1;
+                });
+
+                return Mono.from(testee()
+                    .store(TestMessageId.of(index), Preview.from(String.valueOf(counter))))
+                    .then();
             })
-            .threadCount(threadCount)
-            .operationCount(stepCount)
-            .runSuccessfullyWithin(Duration.ofMinutes(1));
+            .then()
+            .block();
 
-        IntStream.range(0, threadCount * stepCount)
+        IntStream.range(0, stepCount)
             .forEach(index -> assertThat(Mono.from(testee()
                     .retrieve(TestMessageId.of(index)))
                     .block())
-                .isEqualTo(Preview.from(String.valueOf(index))));
+                .isEqualTo(Preview.from(String.valueOf(repeatCount))));
```

Comment:
This assertion is brittle as we do not get the guarranty that index computation (an atomic computation) can not be interleaved with store update (also another atomic operation)

Two atomic operation don't compose into a single atomic operation...

If you really wish to test concurrency, then we need to find something better suited. Can I also suggest doing simpler concurrent test?